graph-builder/src/graph.rs

```diff
@@ -93,28 +113,16 @@ pub fn run(opts: &config::Options, state: &State) -> ! {
     }
 }
 
-fn create_graph(
-    opts: &config::Options,
-    username: Option<&str>,
-    password: Option<&str>,
-    cache: &mut HashMap<u64, Option<registry::Release>>,
+pub fn create_graph(
+    releases: Vec<registry::Release>,
```

Comment:
Few minor doubts on the choice of this `Vec` datastructure: are releases supposed to be unique in this vector? Are release supposed to be order-insensitive in this vector?

If the answer to both of the above is positive, then an `HashSet` is a better choice to encode the invariant.

Reply:
I had the same thoughts and then I went down the rabbit whole, which I just did again to give you the error output :-)

```
error[E0277]: the trait bound `std::collections::HashMap<std::string::String, std::string::String>: std::hash::Hash` is not satisfied
  --> graph-builder/src/release.rs:30:5
   |
30 |     pub metadata: HashMap<String, String>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::hash::Hash` is not implemented for `std::collections::HashMap<std::string::String, std::string::String>`
   |
   = note: required by `std::hash::Hash::hash`
```

I'm happy to use a `HashSet` if you have a proper solution for the above.